source/dub/logging.d

```diff
@@ -0,0 +1,383 @@
+    fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void logErrorTag(T...)(string tag, string fmt, lazy T args) nothrow
+{
+  log(LogLevel.error, false, tag, Color.red, fmt, args);
+}
+
+/**
+  Shorthand function to log a message with error level, using the default
+  tag "Error". The tag color is also fixed to red.
+
+  Params:
+    level = The log level for the logged message
+    fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void logError(T...)(string fmt, lazy T args) nothrow
+{
+  log(LogLevel.error, false, "Error", Color.red, fmt, args);
+}
+
+/**
+  Log a message with the specified log level and with the specified tag string
+  and color. If the log level is debug or diagnostic, the tag is not printed
+  thus the tag string and tag color will be ignored. If the log level is error
+  or warning, the tag will be in bold text. Also the tag can be disabled (for
+  any log level) by passing true as the second argument.
+
+  Params:
+    level = The log level for the logged message
+    disableTag = Setting this to true disables the tag, no matter what
+    tag = The string the tag at the beginning of the line should contain
+    tagColor = The color the tag string should have
+    fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void log(T...)(
+  LogLevel level,
+  bool disableTag,
+  string tag,
+  Color tagColor,
+  string fmt,
+  lazy T args
+) nothrow
+{
+  if (level < _minLevel)
+    return;
+
+  auto hasTag = true;
+  if (level <= LogLevel.diagnostic)
+    hasTag = false;
+  if (disableTag)
+    hasTag = false;
+
+  auto boldTag = false;
+  if (level >= LogLevel.warn)
+    boldTag = true;
+
+  try
+  {
+    string result = format(fmt, args);
+
+    if (hasTag)
+      result = tag.rightJustify(TAG_WIDTH, ' ').color(tagColor, boldTag ? Mode.bold : Mode.init) ~ " " ~ result;
+
+    import dub.internal.colorize : cwrite;
+
+    File output = (level <= LogLevel.info) ? stdout : stderr;
+
+    if (output.isOpen)
+    {
+      output.cwrite(result, "\n");
+      output.flush();
+    }
+  }
+  catch (Exception e)
+  {
+    debug assert(false, e.msg);
+  }
+}
+
+/**
+  Colors the specified string with the specified color. The function is used to
+  print colored text within a log message. The function also checks whether
+  color output is enabled or disabled (when not outputting to a TTY) and, in the
+  last case, just returns the plain string. This allows to use it like so:
+
+  logInfo("Tag", Color.green, "My %s log message", "colored".color(Color.red));
+
+  without worring whether or not colored output is enabled or not.
+
+  Also a mode can be specified, such as bold/underline/etc...
+
+  Params:
+    str = The string to color
+    color = The color to apply
+    mode = An optional mode, such as bold/underline/etc...
+*/
+string color(const string str, const Color color, const Mode mode = Mode.init)
+{
+  import dub.internal.colorize;
+
+  if (_printColors == true)
```

Comment:
This is a weird style, why not `if (_printColors)` ?

Reply:
Right, not sure how that line ended up being that way